### PR TITLE
Async validators + Recaptcha

### PR DESCRIPTION
**Pull request by omarryhan**
#13 

New features:
- Support for async and sync validators
- testing py37

The code I added isn't of best quality, especially the tests. This is only meant to be a prototype.

I think in order to merge this feature, we'd have to go 1 of the following paths:

1. Break backward compatibility and make `validate_on_submit` a coroutine.
2. Keep both `validate_on_submit` and `validate_on_submit_async` (maybe name it differently). However, if we do this, then the user shouldn't call `validate_on_submit_async` followed by `validate_on_submit`, because `validate_on_submit_async` will do lots of monkey patching (We can have some sort of flag though that explicitly enables this functionality).
3. Make a new `SanicForm` class. Maybe call it `SanicForm_` or sth
4. Do nothing :smile:

### CODE REVIEW

**Comment by omarryhan** [comment]
So, I did some cleanup. I think the only thing missing is patching FieldList. If left as is, then it will only be called synchronously by any given form. However, it will fail if any of it's methods are coroutines as they will not be awaited. Basically, as long as you don't use custom FieldList with coroutines in it, you're fine.

As for the tests coverage, I think I covered the main use cases. There might be some untested edge cases though, so if someone can help me spot some, it would be great.

**Comment by pyx** [comment]
Thanks for the feature, I was very busy lately, I did not have time to go through the code right now, do you mind give me sometime.  Thanks again.

**Comment by omarryhan** [comment]
Sure, np! :)

**Comment by pyx** [comment]
Hi Omar, sorry that I was super busy lately, it seems you put a lot of effort into this, I will go through this in a few days.

Thank you so much.

**Comment by pyx** [comment]
Hi Omar Ryhan
I am sorry I failed in manage time to update the project, are you still interested in using this? If you are interested, I will put you into  collaborators list.

**Comment by pyx** [comment]
@mikekeda  How about this one, @omarryhan invested a lot of time into this, what's your thought?

**Comment by pyx** [comment]
@mikekeda I at OP accidentally in last post.

**Comment by omarryhan** [comment]
Hey @pyx. No worries.

I only passively maintain the Sanic projects I was working on. The Sanic team recently introduced some breaking changes to the request context API and I'm not really sure whether my PR might break things or not. My Sanic projects are still running on their latest LTS version which differs from their latest version. I'm pip installing from my own sanic-wtf repo, and things are working fine.

I suggest we leave this PR open until someone shows some interest and maybe tests things out on the latest Sanic version, then maybe we can merge it?

**Comment by omarryhan** [comment]
I must also add that I monkey patched some of wtform's private methods to make this lib support async validations. The downside of this that wtforms cannot guarantee us that they won't break their private APIs. So I hard-coded the wtforms version to avoid things breaking. One downside of that is that we'll not be getting the latest security patches from wtforms should there be any.

**Comment by pyx** [comment]
@omarryhan you are right, relying on private API is fragile to say the least and we should try to avoid. We will wait and see. 

Please remember to add your name in AUTHOR with next commit, not necessarily this one but can be something like a separate one updating python versions tested upon, you have the commit rights.
Thank you.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+matrix:
+  include:
+    - python: 3.7
+      dist: xenial
+      sudo: true
+# </Workaround>
+install:
+  pip install tox-travis
+script: 
+  tox
```
Comment (mikekeda) [inline .travis.yml:15]: add a new line to the end

File: sanic_wtf/__init__.py [hunk 2]
```diff
@@ -7,12 +7,16 @@
 from wtforms.csrf.session import SessionCSRF
 from wtforms.meta import DefaultMeta
 from wtforms.validators import DataRequired, StopValidation
+from wtforms.fields.core import Field
 
-__version__ = '0.6.0.dev0'
+from ._patch import patch
+from .recaptcha import RecaptchaField
```
Comment (mikekeda) [inline sanic_wtf/__init__.py:13]: it's better to use absolute imports

File: sanic_wtf/__init__.py [hunk 3]
```diff
@@ -110,7 +114,6 @@ def getlist(self, name, default=None):
         """
         return super().get(name, default)
 
-
```
Comment (mikekeda) [inline sanic_wtf/__init__.py:113]: according to pep8 here should be 2 blank lines, no need to remove 1 blank line
https://www.python.org/dev/peps/pep-0008/#blank-lines

File: sanic_wtf/__init__.py [hunk 4]
```diff
@@ -141,7 +149,35 @@ def __init__(self, request=None, *args, meta=None, **kwargs):
...
+        if self.request is not None:
+            for name, field in self._fields.items():
+                if hasattr(field, '_get_app'):
+                    field._get_app(self.request.app)
+
+    # @unpatch ??
     def validate_on_submit(self):
+        ''' For async validators: use self.validate_on_submit_async.
+            This method is only still here for backward compatibility
+        '''
```
Comment (mikekeda) [inline sanic_wtf/__init__.py:162]: use double quotes for all docstring
https://www.python.org/dev/peps/pep-0257/#what-is-a-docstring

File: sanic_wtf/_patch.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        If provided, is a dict mapping field names to a sequence of
+        callables which will be passed as extra validators to the field's
+        `validate` method.
+
+    Returns `True` if no errors occur.
+    """
+    self._errors = None
+    success = True
+    for name, field in self._fields.items():
+        if extra_validators is not None and name in extra_validators:
```
Comment (mikekeda) [inline sanic_wtf/_patch.py:40]: it's more pythonic to use `if extra_validators` instead of `extra_validators is not None`
but be aware that for example `if 0` returns False and `if 0 is not None` returns True

File: sanic_wtf/__init__.py [hunk 6]
```diff
@@ -7,12 +7,16 @@
 from wtforms.csrf.session import SessionCSRF
 from wtforms.meta import DefaultMeta
 from wtforms.validators import DataRequired, StopValidation
+from wtforms.fields.core import Field
```
Comment (mikekeda) [inline sanic_wtf/__init__.py:10]: Field isn't used

File: sanic_wtf/_patch.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
+import types
+import inspect
+import itertools
+import functools
+import asyncio
+import warnings
+from wtforms import Form
```
Comment (mikekeda) [inline sanic_wtf/_patch.py:7]: those 3 imports aren't used

File: sanic_wtf/_patch.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    """
+    extra = {}
+    for name in self._fields:
+        inline = getattr(self.__class__, "validate_%s" % name, None)
+        if inline is not None:
+            extra[name] = [inline]
+
+    # Added an await here
+    return await self.validate_base(extra)
+
```
Comment (mikekeda) [inline sanic_wtf/_patch.py:25]: add 1 blank line

File: sanic_wtf/_patch.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    Subfields shouldn't override this, but rather override either
+    `pre_validate`, `post_validate` or both, depending on needs.
+
+    :param form: The form the field belongs to.
+    :param extra_validators: A sequence of extra validators to run.
+    """
+    self.errors = list(self.process_errors)
+    stop_validation = False
+
+
```
Comment (mikekeda) [inline sanic_wtf/_patch.py:65]: remove blank line

File: sanic_wtf/_patch.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    try:
+        # Await if a coroutine
+        res = self.post_validate(form, stop_validation)
+        if inspect.isawaitable(res):
+            await res
+    except ValueError as e:
+        self.errors.append(e.args[0])
+
+    return len(self.errors) == 0
+
```
Comment (mikekeda) [inline sanic_wtf/_patch.py:94]: add 1 blank line

File: sanic_wtf/_patch.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+            if inspect.isawaitable(res):
+                await res
+        except StopValidation as e:
+            if e.args and e.args[0]:
+                self.errors.append(e.args[0])
+            return True
+        except ValueError as e:  # Catches validation errors
+            self.errors.append(e.args[0])
+    return False    
+
```
Comment (mikekeda) [inline sanic_wtf/_patch.py:117]: add 1 blank line

File: sanic_wtf/_patch.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+                if hasattr(field, '_run_validation_chain'):
+                    setattr(
+                        self._fields[field_name],
+                        '_run_validation_chain',
+                        types.MethodType(_run_validation_chain_async, self._fields[field_name])
+                    )
+            # # 3.2 Patch fieldlist type
+            # # Won't work asynchronously until "entries" (subfields) are dealt with
+        self.patched = True
+
```
Comment (mikekeda) [inline sanic_wtf/_patch.py:158]: add 1 blank line

File: sanic_wtf/_patch.py [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+def patch(f):
+    ''' Patch wtforms to make it support async validators
```
Comment (mikekeda) [inline sanic_wtf/_patch.py:160]: use double quotes to docstrings

File: sanic_wtf/_patch.py [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    self.errors = list(self.process_errors)
+    stop_validation = False
+
+
+    # Call pre_validate
+    try:
+        # Await if a coroutine
+        res = self.pre_validate(form)
+        if inspect.isawaitable(res):
+            res = await(res)
```
Comment (mikekeda) [inline sanic_wtf/_patch.py:71]: res isn't used,
`await res` looks better to me compare to `await(res) `

File: sanic_wtf/recaptcha.py [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+
+from wtforms.fields import Field
+from wtforms import ValidationError
+
+import aiorecaptcha
+from markupsafe import Markup
+
+
+__all__ = ['RecaptchaField']
+
```
Comment (mikekeda) [inline sanic_wtf/recaptcha.py:13]: add 1 blank line

File: sanic_wtf/recaptcha.py [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+    # Verify
+    try:
+        await aiorecaptcha.verify(
+            secret=config[field._config_prefix + '_PRIVATE_KEY'],
+            response=response,
+            remoteip=ip
+        )
+    except aiorecaptcha.RecaptchaError as e:
+        raise ValidationError(e.msg)
+
```
Comment (mikekeda) [inline sanic_wtf/recaptcha.py:48]: add 1 blank line

File: sanic_wtf/recaptcha.py [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+        badge=field._config.get(self._config_prefix + '_BADGE'),
+        size=field._config.get(self._config_prefix + '_SIZE'),
+        type_=field._config.get(self._config_prefix + '_TYPE'),
+        tabindex=field._config.get(self._config_prefix + '_TABINDEX'),
+        callback=field._config.get(self._config_prefix + '_CALLBACK'),
+        expired_callback=field._config.get(self._config_prefix + '_EXPIRED_CALLBACK'),
+        error_callback=field._config.get(self._config_prefix + '_ERROR_CALLBACK')
+    )
+    return Markup(js) + Markup(html)
+
```
Comment (mikekeda) [inline sanic_wtf/recaptcha.py:71]: add 1 blank line

File: sanic_wtf/recaptcha.py [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+class RecaptchaField(Field):
+    '''
```
Comment (mikekeda) [inline sanic_wtf/recaptcha.py:73]: use double quotes for docstrings

File: tests/conftest.py [hunk 19]
```diff
@@ -15,3 +19,35 @@ async def add_session(request):
         request[name] = session
 
     return test_app
+
```
Comment (mikekeda) [inline tests/conftest.py:22]: add 1 blank line before each fixture

File: tests/helpers.py [hunk 20]
```diff
@@ -0,0 +1,12 @@
+# NOTE
+# taking shortcut here, assuming there will be only one "string" (the token)
+# ever longer than 40.
+csrf_token_pattern = '''value="([0-9a-f#]{40,})"'''
+
```
Comment (mikekeda) [inline tests/helpers.py:5]: add 1 blank line

File: tests/test_async_validators.py [hunk 21]
```diff
@@ -0,0 +1,222 @@
+import re
+import asyncio
```
Comment (mikekeda) [inline tests/test_async_validators.py:2]: asyncio isn't used

File: tests/test_async_validators.py [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+
+from sanic import response
+from wtforms.validators import DataRequired, Length, ValidationError
```
Comment (mikekeda) [inline tests/test_async_validators.py:5]: ValidationError isn't used

File: tests/test_async_validators.py [hunk 23]
```diff
@@ -0,0 +1,222 @@
...
+from wtforms import FileField, StringField, SubmitField
```
Comment (mikekeda) [inline tests/test_async_validators.py:6]: FileField isn't used

File: tests/test_async_validators.py [hunk 24]
```diff
@@ -0,0 +1,222 @@
...
+
+from sanic_wtf import SanicForm, to_bytes
```
Comment (mikekeda) [inline tests/test_async_validators.py:8]: to_bytes isn't used

File: tests/test_async_validators.py [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+    assert resp.status == 200
+    assert 'csrf_token' in resp.text
+    token = re.findall(csrf_token_pattern, resp.text)[0]
+    assert token
+
+    payload = {'msg': 'happy', 'csrf_token': token}
+    req, resp = app.test_client.post('/', data=payload)
+    assert resp.status == 200
+    assert 'valid' in resp.text
+
```
Comment (mikekeda) [inline tests/test_async_validators.py:52]: add 1 blank line before each test

File: tests/test_form.py [hunk 26]
```diff
@@ -1,27 +1,14 @@
 # -*- coding: utf-8 -*-
 import re
 import os.path
+import asyncio
```
Comment (mikekeda) [inline tests/test_form.py:4]: asyncio isn't used

File: tests/test_form.py [hunk 27]
```diff
@@ -1,27 +1,14 @@
...
 
 from sanic import response
-from wtforms.validators import DataRequired, Length
+from wtforms.validators import DataRequired, Length, ValidationError
```
Comment (mikekeda) [inline tests/test_form.py:7]: ValidationError isn't used

File: tests/test_form.py [hunk 28]
```diff
@@ -173,7 +161,6 @@ async def index(request):
     assert resp.status == 200
     assert 'validated' in resp.text
 
-
```
Comment (mikekeda) [inline tests/test_form.py:176]: why?
here should be 2 blank lines according to pep8 https://www.python.org/dev/peps/pep-0008/#blank-lines

File: tests/test_patching.py [hunk 29]
```diff
@@ -0,0 +1,191 @@
+import asyncio
+
+import pytest
+from wtforms import Form, Field, FieldList, StringField
```
Comment (mikekeda) [inline tests/test_patching.py:4]: Form, Field, FieldList aren't used

File: tests/test_patching.py [hunk 30]
```diff
@@ -0,0 +1,191 @@
...
+
+    1. wtforms.Form.validate() --> _wtforms_form_validate
+    2. wtforms.BaseForm.validate() --> _wtforms_base_form_validate
+    3. fields and field lists
+    3.1 wtforms.fields.core.Field
+    3.1.1. wtforms.fields.core.Field.validate() --> _field_validate
+    3.1.2. wtforms.fields.core.Field._run_validation_chain() --> _run_validation_chain_async 
+    TODO: 3.2 wtforms.fields.core.FieldList
+'''
+
```
Comment (mikekeda) [inline tests/test_patching.py:24]: add 1 blank line before each function

File: tests/test_patching.py [hunk 31]
```diff
@@ -0,0 +1,191 @@
...
+    form = _Form()
+    assert hasattr(form, 'validate_base') is False
+    _patch(form)
+    assert asyncio.iscoroutinefunction(form.validate_base)
+
+# test validate methods
+def test_all_fields_have_async_validate_method():
+
+    async def async_val(form, field):
```
Comment (mikekeda) [inline tests/test_patching.py:56]: `form, field` aren't used, could we use `_, __` instead?

File: tests/test_patching.py [hunk 32]
```diff
@@ -0,0 +1,191 @@
...
+            ) is False
+    _patch(form)
+    for _, field in form._fields.items():
+        if hasattr(field, 'validate'):
+            assert asyncio.iscoroutinefunction(
+                field.validate
+            ) is True
+
+# Pre/Post Validators
+
```
Comment (mikekeda) [inline tests/test_patching.py:82]: add 1 blank line

File: tests/test_patching.py [hunk 33]
```diff
@@ -0,0 +1,191 @@
...
+class AsyncValidatorFound(Exception):
+    pass
+
+@pytest.fixture
+def req():
+    class App:
+        pass
+    class Request:
```
Comment (mikekeda) [inline tests/test_patching.py:90]: add 1 blank line before

File: tests/test_patching.py [hunk 34]
```diff
@@ -0,0 +1,191 @@
...
+            self.files = None
+    req_ = Request()
+    req_.method = 'POST'
+    return req_
+
+
+@pytest.mark.asyncio
+async def test_post_validator_field_patch(req):
+    POST_VAL_ERR_MSG = 'post_patched_and_called'
+    class AsyncPostValidatorField(StringField):
```
Comment (mikekeda) [inline tests/test_patching.py:103]: add 1 blank line before

File: tests/test_patching.py [hunk 35]
```diff
@@ -0,0 +1,191 @@
...
+        req
+    )
+    with pytest.raises(AsyncValidatorFound) as e:
+        await form.validate_on_submit_async()
+        assert POST_VAL_ERR_MSG in str(e)
+    
+@pytest.mark.asyncio
+async def test_post_validator_field_sync(req):
+    POST_VAL_ERR_MSG = 'post_patched_and_called'
+    class AsyncPostValidatorField(StringField):
```
Comment (mikekeda) [inline tests/test_patching.py:121]: add 1 blank line before

File: tests/test_patching.py [hunk 36]
```diff
@@ -0,0 +1,191 @@
...
+    )
+    assert await form.validate_on_submit_async()
+
+@pytest.mark.asyncio
+async def test_post_validator_field_async_fails(req):
+    class AsyncPostValidatorField(StringField):
+        async def post_validate(self, form, stop_validation):
+            pass
+
+    async def fail(*args):
```
Comment (mikekeda) [inline tests/test_patching.py:157]: `*args` isn't used, could we use `*_` instead?

File: tests/test_patching.py [hunk 37]
```diff
@@ -0,0 +1,191 @@
...
+
+    req.form = CIMultiDict(dict(string='asdasdsda'))
+    form = _Form(
+        req
+    )
+
+    assert await form.validate_on_submit_async() is False
+    assert 'I failed' in str(form.errors)
+    assert 'I failed again' in str(form.errors)
+
```
Comment (mikekeda) [inline tests/test_patching.py:191]: remove spaces

File: tests/test_recaptcha.py [hunk 38]
```diff
@@ -0,0 +1,154 @@
+import asyncio
+
+import pytest
+from wtforms import ValidationError
```
Comment (mikekeda) [inline tests/test_recaptcha.py:4]: ValidationError isn't used

File: tests/test_recaptcha.py [hunk 39]
```diff
@@ -0,0 +1,154 @@
...
+    
+    class App:
+        def __init__(self, config):
+            self.config = config
+
+
+    _base_config = {'WTF_CSRF_ENABLED': False}
+    test_app = App({**_base_config, **config})
+    return Request(test_app)
+
```
Comment (mikekeda) [inline tests/test_recaptcha.py:27]: add 1 blank line before each top level function
(here should be 2 blank lines in total - https://www.python.org/dev/peps/pep-0008/#blank-lines)

File: tests/test_recaptcha.py [hunk 40]
```diff
@@ -0,0 +1,154 @@
...
+            self.files = None
+            self.form = multidict.CIMultiDict()
+            self.json = multidict.CIMultiDict()
+            self.method = 'POST' 
+    
+    class App:
+        def __init__(self, config):
+            self.config = config
+
+
```
Comment (mikekeda) [inline tests/test_recaptcha.py:23]: remove 1 blank line

File: sanic_wtf/recaptcha.py [hunk 41]
```diff
@@ -0,0 +1,239 @@
...
+
+                * Default False
+
+                * You might need this if you only want to use Recaptcha's JS script (Recaptcha V3)
+    '''
+
+    widget = recaptcha_widget
+            
+    def __init__(self, label='Recaptcha', extra_validators: Union[list, tuple, set]=None, **kwargs):
+        validators = set([recaptcha_validator])
```
Comment (mikekeda) [inline sanic_wtf/recaptcha.py:232]: `validators = {recaptcha_validator}`
will be more pythonic

File: sanic_wtf/recaptcha.py [hunk 42]
```diff
@@ -0,0 +1,239 @@
...
+            RECAPTCHA_JS_ONLY (bool):
+
+                * Default False
+
+                * You might need this if you only want to use Recaptcha's JS script (Recaptcha V3)
+    '''
+
+    widget = recaptcha_widget
+            
+    def __init__(self, label='Recaptcha', extra_validators: Union[list, tuple, set]=None, **kwargs):
```
Comment (mikekeda) [inline sanic_wtf/recaptcha.py:231]: add spaces before and after `=` if you use typing
should be `extra_validators: Union[list, tuple, set] = None` instead of `extra_validators: Union[list, tuple, set]=None`

File: tox.ini [hunk 43]
```diff
@@ -1,6 +1,9 @@
 [tox]
-envlist = py35, py36
+envlist = py36, py37
 
 [testenv]
+passenv = py.test clear_pycache.sh python python3.7 python3.6 PYTHONPATH
 deps = -rrequirements-dev.txt
-commands = py.test
+commands = 
+    ./clear_pycache.sh
```
Comment (mikekeda) [inline tox.ini:8]: why do we nee to clear pycache?

**Review by mikekeda — CHANGES_REQUESTED** [review]

**Comment by mikekeda** [comment]
> @mikekeda How about this one, @omarryhan invested a lot of time into this, what's your thought?

1. I think it's better to split this into 2 RPs Async validators and Recaptcha
2. monkey patching doesn't look good to me, can we create a separate class for this?
3. What performance improvement we will get with Async validators?
4. I don't think it's a good idea to make RECAPTCHA required, it's better to make it optional (in `RECAPTCHA_PUBLIC_KEY ` and `RECAPTCHA_PRIVATE_KEY ` are marked as required)
5. Why `clear` and `clear_pycache.sh` are committed?

**Comment by omarryhan** [comment]
Thanks for taking the time to review @mikekeda 

I added async not for the performance gain but to be able to call async code for validations, like querying a database to check if an email exists.

I'll address these issues once people show enough interest in the PR. Right now, it seems like I'm the only one using it.
